File: android/app/src/main/java/com/example/smartplanner/data/ProductivityStatsService.kt
package com.example.smartplanner.data

import com.example.smartplanner.model.ProductivityStats
import com.example.smartplanner.model.Task
import com.example.smartplanner.model.TaskPriority
import kotlin.math.roundToInt

class ProductivityStatsService {
    fun calculate(tasks: List<Task>): ProductivityStats {
        val totalTasks = tasks.size
        val completedTasks = tasks.count { it.done }
        val totalSubtasks = tasks.sumOf { it.subtasks.size }
        val completedSubtasks = tasks.sumOf { task -> task.subtasks.count { it.done } }

        return ProductivityStats(
            totalTasks = totalTasks,
            completedTasks = completedTasks,
            activeTasks = totalTasks - completedTasks,
            completionPercent = completionPercent(completedTasks, totalTasks),
            highPriorityTasks = tasks.count { it.priority == TaskPriority.HIGH },
            flaggedTasks = tasks.count { it.flagged },
            overdueOrDeadlineTasks = tasks.count { it.deadlineText != null },
            totalSubtasks = totalSubtasks,
            completedSubtasks = completedSubtasks
        )
    }

    private fun completionPercent(completed: Int, total: Int): Int {
        if (total == 0) return 0
        return ((completed.toDouble() / total.toDouble()) * 100).roundToInt()
    }
}

File: ios/SmartPlanner/Services/ProductivityStatsService.swift
import Foundation

struct ProductivityStatsService {
    func calculate(tasks: [TaskModel]) -> ProductivityStats {
        let totalTasks = tasks.count
        let completedTasks = tasks.filter(\.isDone).count
        let totalSubtasks = tasks.reduce(0) { $0 + $1.subtasks.count }
        let completedSubtasks = tasks.reduce(0) { total, task in
            total + task.subtasks.filter(\.isDone).count
        }

        return ProductivityStats(
            totalTasks: totalTasks,
            completedTasks: completedTasks,
            activeTasks: totalTasks - completedTasks,
            completionPercent: completionPercent(completed: completedTasks, total: totalTasks),
            highPriorityTasks: tasks.filter { $0.priority == .high }.count,
            flaggedTasks: tasks.filter(\.isFlagged).count,
            deadlineTasks: tasks.filter(\.hasDeadline).count,
            totalSubtasks: totalSubtasks,
            completedSubtasks: completedSubtasks
        )
    }

    private func completionPercent(completed: Int, total: Int) -> Int {
        guard total > 0 else { return 0 }
        return Int((Double(completed) / Double(total) * 100).rounded())
    }
}
